fido2/tpm.py

```diff
@@ -0,0 +1,308 @@
+# POSSIBILITY OF SUCH DAMAGE.
+
+import struct
+from io import BytesIO
+from binascii import b2a_hex
+
+from enum import IntEnum, IntFlag
+from cryptography.hazmat.backends import default_backend
+from cryptography.hazmat.primitives.asymmetric import rsa
+
+from .utils import bytes2int
+
+
+TPM_ALG_NULL = 0x0010
+
+
+class TpmAlgAsym(IntEnum):
+    RSA = 0x0001
+    ECC = 0x0023
+
+
+class TpmAlgHash(IntEnum):
+    SHA1 = 0x0004
+    SHA256 = 0x000B
+    SHA384 = 0x000C
+    SHA512 = 0x000D
+
+
+class TpmAttestationFormat(object):
+    """the signature data is defined by [TPMv2-Part2] Section 10.12.8 (TPMS_ATTEST)
+    as:
+      TPM_GENERATED_VALUE (0xff544347 aka "\xffTCG")
+      TPMI_ST_ATTEST - always TPM_ST_ATTEST_CERTIFY (0x8017)
+        because signing procedure defines it should call TPM_Certify
+        [TPMv2-Part3] Section 18.2
+      TPM2B_NAME
+        size (uint16)
+        name (size long)
+      TPM2B_DATA
+        size (uint16)
+        name (size long)
+      TPMS_CLOCK_INFO
+        clock (uint64)
+        resetCount (uint32)
+        restartCount (uint32)
+        safe (byte) 1 yes, 0 no
+      firmwareVersion uint64
+      attested TPMS_CERTIFY_INFO (because TPM_ST_ATTEST_CERTIFY)
+        name TPM2B_NAME
+        qualified_name TPM2B_NAME
+    See:
+      https://www.trustedcomputinggroup.org/wp-content/uploads/TPM-Rev-2.0-Part-2-Structures-01.38.pdf
+      https://www.trustedcomputinggroup.org/wp-content/uploads/TPM-Rev-2.0-Part-3-Commands-01.38.pdf
+    """
+
+    TPM_GENERATED_VALUE = b"\xffTCG"
+    TPM_ST_ATTEST_CERTIFY = b"\x80\x17"
+
+    @classmethod
+    def parse(cls, data):
+        reader = BytesIO(data)
+        generated_value = reader.read(4)
+
+        if generated_value != cls.TPM_GENERATED_VALUE:
+            raise ValueError("generated value field is invalid")
+
+        tpmi_st_attest = reader.read(2)
+        if tpmi_st_attest != cls.TPM_ST_ATTEST_CERTIFY:
+            raise ValueError("tpmi_st_attest field is invalid")
+
+        try:
+            name_len = struct.unpack("!H", reader.read(2))[0]
+            name = reader.read(name_len)
+            if name_len != len(name):
+                raise ValueError("name is too short")
+
+            data_len = struct.unpack("!H", reader.read(2))[0]
+            data = reader.read(data_len)
+            if data_len != len(data):
+                raise ValueError("data is too short")
+
+            clock = struct.unpack("!Q", reader.read(8))[0]
+            reset_count = struct.unpack("!L", reader.read(4))[0]
+            restart_count = struct.unpack("!L", reader.read(4))[0]
+            safe_value = reader.read(1)
+            if safe_value not in (b"\x00", b"\x01"):
+                raise ValueError(
+                    "invalid value 0x{} for boolean".format(
+                        b2a_hex(safe_value).decode("ascii")
+                    )
+                )
+            safe = safe_value == "\x01"
```

Comment:
This looks like a bug, comparing bytes to string?

Reply:
Good catch, could probably be rewritten with the wrapper you mentioned below. Fixed up for now.